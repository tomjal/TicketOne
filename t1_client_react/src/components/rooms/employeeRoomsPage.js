import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { getAllOpenRoomsIds } from './../../actions/roomsActionCreators';
import { getAllMessages, sendMessageToRoom } from './../../actions/messagesActionCreators';
import { getEmployeeInitRoomsData } from './../../actions/combinedActionCreators';

import RoomWidget from './roomWidget';

const MAX_FLEX_ELEMS_IN_ROW = 3;

const mapStateToProps = (state) => {
  return {
    id: state.context.id,
    role: state.context.role,
    rooms: state.rooms,
    messages: state.messages
  }
};

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators({
    getAllOpenRoomsIds,
    getAllMessages,
    sendMessageToRoom,
    getEmployeeInitRoomsData
  }, dispatch);
};

export class EmployeeRoomsPage extends Component {
  constructor(props) {
    super(props)
    this.sendMessageCallback = this.sendMessageCallback.bind(this);
  }
  componentDidMount() {
    //combined init data employee
    this.props.getEmployeeInitRoomsData();
  }
  sendMessageCallback(message, id) {
    this.props.sendMessageToRoom(message, id, this.props.role, this.props.id);
  }
  getEmptyFlexFillers(numbInRow, numbOfAll) {
    let emptyElemsNumber = numbInRow - (numbOfAll % numbInRow);
    let emptyFlexElems = [];
    for (let i = 0; i < emptyElemsNumber; i++) {
      emptyFlexElems.push({ id: i });
    }
    return emptyFlexElems;
  }
  render() {
    const { role, rooms, messages } = this.props;
    const emptyFlexElems = this.getEmptyFlexFillers(MAX_FLEX_ELEMS_IN_ROW, rooms.length);
    return (
      <div>
        {rooms.length !== 0 &&
          <div className="room-flex-container">
            {rooms.map((room, i) => {
              let localMessages = [];
              if (messages[room.id]) {
                localMessages = messages[room.id];
              }
              return <RoomWidget key={i}
                topic={room.roomTopic}
                isOpened={room.isOpened}
                id={room.id}
                creatorId={room.clientId}
                role={role}
                sendMessageCallback={this.sendMessageCallback}
                messages={localMessages} />;
            })}
            {emptyFlexElems.map((roomId, i) => {
              return <div key={`filler-${i}`} className="room-flex-container__filler"></div>;
            })}
          </div>
        }
        {rooms.length === 0 &&
          <div style={{ fontSize: 16 }}>
            There are no open chats currently
          </div>
        }
      </div>
    );
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EmployeeRoomsPage);
