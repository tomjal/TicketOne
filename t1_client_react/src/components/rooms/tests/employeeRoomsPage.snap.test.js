import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { EmployeeRoomsPage } from './../employeeRoomsPage';

test('EmployeeRoomsPage snapshot', () => {
  const props = {
    role: "",
    messages: [],
    rooms: [],
    stats: { solved: 0, unsolved: 0 },
    getEmployeeInitRoomsData: jest.fn(),
    getGlobalSolvedUnsolvedStatistics: jest.fn(),
    sendMessageToRoom: jest.fn()
  };

  const wrap = shallow(<EmployeeRoomsPage {...props} />);

  expect(shallowToJson(wrap)).toMatchSnapshot();
});