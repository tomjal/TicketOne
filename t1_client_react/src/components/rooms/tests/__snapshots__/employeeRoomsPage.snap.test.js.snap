// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeRoomsPage snapshot 1`] = `
<div>
  <StatsWidget
    statsObject={
      Object {
        "solved": 0,
        "unsolved": 0,
      }
    }
  />
  <div
    style={
      Object {
        "fontSize": 16,
      }
    }
  >
    There are no open chats currently
  </div>
</div>
`;
